improve exception messages used within import dialog

cosmicexplorer opened the pull request:
- Provide better exception messages when `PantsUtil.listAllTargets()` fails.
  - e.g. whether the process finished with an exit code or was interrupted
- Log the exit code and output file if the process ran to completion.

Currently, the function throws a `PantsException` on failure which describes the command line that failed. This command line already contains the output file, incidentally, but this change makes the output file explicit.

cosmicexplorer commented:
CI failed because of the md5sum, just uploaded another commit to fix that. Since `.listAllTargets()` is called exactly once in an integration test, I want to add some unit tests that actually ensure these exceptions are thrown.

wisechengyi commented on common/com/twitter/intellij/pants/util/PantsUtil.java:
```diff
@@ -320,21 +320,34 @@ public static GeneralCommandLine defaultCommandLine(@NotNull File pantsExecutabl
+                        processOutput.getExitCode()) +
+          String.format("argv: '%s'\n", cmd.getCommandLineString()) +
+          String.format("stdout:\n%s\n", listOutput) +
+          String.format("warnings:\n%s\n", processOutput.getStdout()) +
```
if the command fails, nothing will be in `listOutput`.

so stdout will be `processOutput.getStdout()` and stderr will be `processOutput.getStderr()`

cosmicexplorer replied:
That makes sense -- I've made the exception message contain just stdout and stderr as you described.

wisechengyi commented on common/com/twitter/intellij/pants/util/PantsUtil.java:
```diff
@@ -320,21 +320,34 @@ public static GeneralCommandLine defaultCommandLine(@NotNull File pantsExecutabl
+          String.format("argv: '%s'\n", cmd.getCommandLineString()) +
+          String.format("stdout:\n%s\n", listOutput) +
+          String.format("warnings:\n%s\n", processOutput.getStdout()) +
+          String.format("stderr:\n%s\n", processOutput.getStderr());
```
nit: instead of `\n` on every line, put them in an array then join with `\n`

cosmicexplorer replied:
Fixed.

wisechengyi commented on common/com/twitter/intellij/pants/util/PantsUtil.java:
```diff
@@ -320,21 +320,34 @@ public static GeneralCommandLine defaultCommandLine(@NotNull File pantsExecutabl
       else {
-        throw new PantsException("Failed:" + cmd.getCommandLineString());
+        final String errorDescription =
+          String.format("could not list targets: pants exited with status %d\n",
```
nit: capitalization `Could not...... Pants ...`

cosmicexplorer replied:
Fixed, also fixed in the other exception message to `"Could not execute command..."`.

wisechengyi commented on common/com/twitter/intellij/pants/util/PantsUtil.java:
```diff
@@ -320,21 +320,34 @@ public static GeneralCommandLine defaultCommandLine(@NotNull File pantsExecutabl
+        String.format("could not execute command: '%s' due to error: '%s'",
+                      cmd.getCommandLineString(),
+                      e.getMessage());
+      LOG.warn(noProcessErrMsg, e);
```
nit: 
1.`noProcessErrMsg` may suggests boolean, maybe `otherProcessErrMsg`
2. capitalization in error message

cosmicexplorer replied:
Changed to `processCreationFailureMessage` and fixed capitalization.

cosmicexplorer commented:
I removed `.loadTargets()` and added the build file path check to `PantsUtil.listAllTargets()`, which allowed me to test both the `PantsUtil` function as well as the UI action. This and the previous commit both pass CI.

I still don't know how to make a test checking whether the project import dialog does actually get closed after `./pants list ...` fails. There may be some hook that runs when the dialog is dismissed, but I'm not sure. I think that's a lot of effort to do in this pull request, but I'll try to figure out ways to do that in the future.

wisechengyi commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +41,57 @@ public void run() {
+
+  public void testListTargetInvalidBuildFile() {
+    String projectDir = getProjectFolder().getPath();
+    Path invalidBuildFilePath = Paths.get(
```
can probably use something like https://github.com/wisechengyi/intellij-pants-plugin/blob/540e81a0b001aac8bbdef23dd966804d60cc3f4f/testFramework/com/twitter/intellij/pants/testFramework/PantsTestUtils.java#L16 to get testData

cosmicexplorer replied:
Gotcha, I'll try to make this cleaner.

wisechengyi commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +41,57 @@ public void run() {
+    } catch (PantsException e) {
+      caught = true;
+    }
+    assertTrue("PantsException not thrown for invalid BUILD file", caught);
```
you can put the assertion/fail message in the try block, e.g. https://github.com/wisechengyi/intellij-pants-plugin/blob/e01fd6db7c00636108cb95f12bb8bef7fc853e59/tests/com/twitter/intellij/pants/model/PantsOptionsTest.java#L53-L61

so no need for the `caught` variable.

cosmicexplorer replied:
Thanks for the example, I hadn't thought of that. I'll make the change.

cosmicexplorer commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +41,57 @@ public void run() {
+    File projectDir = getProjectFolder();
+    VirtualFile virtualProjectDir = LocalFileSystem.getInstance().findFileByIoFile(projectDir);
+    Optional<VirtualFile> pantsIniFileResult = PantsUtil.findPantsIniFile(Optional.ofNullable(virtualProjectDir));
+    assertTrue(pantsIniFileResult.isPresent());
```
Done, here and elsewhere.

cosmicexplorer commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +41,57 @@ public void run() {
+    String pantsIniFilePath = pantsIniFileResult.get().getPath();
+    File pantsIniFile = new File(pantsIniFilePath);
+    assertTrue(pantsIniFile.exists() && !pantsIniFile.isDirectory());
+    assertTrue(!PantsUtil.isBUILDFilePath(pantsIniFilePath));
```
Done, here and elsewhere.

cosmicexplorer commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +41,57 @@ public void run() {
+    File nonexistentFile = new File(nonexistentFilePath);
+    assertTrue(!nonexistentFile.exists());
+    assertTrue(!PantsUtil.isBUILDFilePath(nonexistentFilePath));
+    assertEquals(PantsUtil.listAllTargets(nonexistentFilePath),
```
That makes sense. I'll fix that in this file and in `PantsProjectSettingsTest.java`.

wisechengyi commented on tests/com/twitter/intellij/pants/util/PantsUtilTest.java:
```diff
@@ -36,4 +38,40 @@ public void run() {
+
+    try {
+      PantsUtil.listAllTargets(invalidBuildFilePath);
+      fail(String.format("%s should have been thrown"));
```
is there a particular exception to be caught? what is %s for?

cosmicexplorer replied:
Whoops. I used `String.format()` following the example you gave in your previous comment, but forgot that means I need to check it. Just pushed a change fixing this.
